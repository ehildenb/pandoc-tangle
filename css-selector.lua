IDSTART  = '%.'
ID       = '[%a%d-]+'
NOTSTART = ':not%('
NOTEND   = '%)'
AND      = '%s+'
OR       = '%s*,%s*'

tokens = { IDSTART , ID , NOTSTART , NOTEND , OR , AND }

function matches_token(input, token)
    return string.match(input, '^' .. token, 1)
end

function tokenize(selector_string)
    local tokenized_input = {}
    local found_token = false
    local match = ""

    while string.len(selector_string) > 0 do
        --- print("current string: " .. selector_string)
        found_token = false
        for _,token in pairs(tokens) do
            --- print("matching token: " .. token)
            match = matches_token(selector_string, token)
            if match then
                --- print("token matched!")
                found_token = true
                selector_string,_ = string.gsub(selector_string, token, '', 1)
                table.insert(tokenized_input, match)
                break
            end
        end
        if not found_token then
            error("Could not tokenize: " .. selector_string)
        end
    end
    return tokenized_input
end

function group_tokens(tokenized)
    local grouped_tokens = {}
    local current_token = 1
    while tokenized[current_token] do
        if matches_token(tokenized[current_token], IDSTART) and matches_token(tokenized[current_token + 1], ID) then
            table.insert(grouped_tokens, { id = tokenized[current_token + 1] })
            current_token = current_token + 2
        elseif matches_token(tokenized[current_token], NOTSTART) then
            local not_group = {}
            current_token = current_token + 1
            while not matches_token(tokenized[current_token], NOTEND) do
                table.insert(not_group, tokenized[current_token])
                current_token = current_token + 1
            end
            current_token = current_token + 1
            table.insert(grouped_tokens, { notExp = group_tokens(not_group) })
        elseif matches_token(tokenized[current_token], OR) then
            table.insert(grouped_tokens, OR)
            current_token = current_token + 1
        elseif matches_token(tokenized[current_token], AND) then
            current_token = current_token + 1
        else
            error("Unexpected token: " .. tokenized[current_token], 1)
        end
    end
    return grouped_tokens
end

function parse_tokens(tokenized)
    return tokenized
end
