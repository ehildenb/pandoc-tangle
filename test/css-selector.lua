require "../css-selector"

--- Test triples
--- ------------

tests = {}

tests[1] = { input  = ".test .input"
           , tokens = { "." , "test" , " " , "." , "input" }
           , groups = { { id = "test" } , { id = "input" } }
           , ast    = { orExp = { { andExp = { { id = "test" } , { id = "input" } } } } }
           , eval   = { { tset = { }                            , result = false }
                      , { tset = { "test" }                     , result = false }
                      , { tset = { "test" , "input" }           , result = true  }
                      , { tset = { "test" , "input" , "extra" } , result = true  }
                      , { tset = { "extra" }                    , result = false }
                      }
           }

tests[2] = { input  = ".test , .input"
           , tokens = { "." , "test"  , " , " , "." , "input" }
           , groups = { { id = "test" } , OR , { id = "input" } }
           , ast    = { orExp = { { andExp = { { id = "input" } } } , { andExp = { { id = "test" } } } } }
           , eval   = { { tset = { }                            , result = false }
                      , { tset = { "test" }                     , result = true  }
                      , { tset = { "test" , "input" }           , result = true  }
                      , { tset = { "test" , "input" , "extra" } , result = true  }
                      , { tset = { "extra" }                    , result = false }
                      }
           }

tests[3] = { input  = ".k.transferFrom-then-branch:not(.transferFrom-else-branch)"
           , tokens = { "." , "k" , "." , "transferFrom-then-branch" , ":not(" , "." , "transferFrom-else-branch" , ")" }
           , groups = { { id = "k" } , { id = "transferFrom-then-branch" } , { notExp = { { id = "transferFrom-else-branch" } } } }
           , ast    = { orExp = { { andExp = { { id = "k" } , { id = "transferFrom-then-branch" } , { notExp = { orExp = { { andExp = { { id = "transferFrom-else-branch" } } } } } } } } } }
           , eval   = { { tset = { } , result = false }
                      , { tset = { "k" }                                                           , result = false }
                      , { tset = { "k" , "transferFrom-then-branch" }                              , result = true  }
                      , { tset = { "k" , "transferFrom-then-branch" , "transferFrom-else-branch" } , result = false }
                      , { tset = { "k" , "transferFrom-else-branch" }                              , result = false }
                      }
           }

tests[4] = { input  = ".k .blah :not(.foo.bar)"
           , tokens = { "." , "k" , " " , "." , "blah" , " " , ":not(" , "." , "foo" , "." , "bar" , ")" }
           , groups = { { id = "k" } , { id = "blah" } , { notExp = { { id = "foo" } , { id = "bar" } } } }
           , ast    = { orExp = { { andExp = { { id = "k" } , { id = "blah" } , { notExp = { orExp = { { andExp = { { id = "foo" } , { id = "bar" } } } } } } } } } }
           , eval   = { { tset = { }                              , result = false }
                      , { tset = { "k" , "blah" }                 , result = true  }
                      , { tset = { "k" , "blah" , "foo" }         , result = true  }
                      , { tset = { "k" , "blah" , "bar" }         , result = true  }
                      , { tset = { "k" , "blah" , "foo" , "bar" } , result = false }
                      , { tset = { "k" , "foo" }                  , result = false }
                      }
           }

tests[5] = { input  = ".k , .rvk , :not(.uiuck)"
           , tokens = { "." , "k" , " , " , "." , "rvk" , " , " , ":not(" , "." , "uiuck" , ")" }
           , groups = { { id = "k" } , OR , { id = "rvk" } , OR , { notExp = { { id = "uiuck" } } } }
           , ast    = { orExp = { { andExp = { { notExp = { orExp = { { andExp = { { id = "uiuck" } } } } } } } } , { andExp = { { id = "rvk" } } } , { andExp = { { id = "k" } } } } }
           , eval   = { { tset = { }                 , result = true  }
                      , { tset = { "k" }             , result = true  }
                      , { tset = { "rvk" }           , result = true  }
                      , { tset = { "extra" }         , result = true  }
                      , { tset = { "uiuck" }         , result = false }
                      , { tset = { "rvk" , "uiuck" } , result = true  }
                      }
           }

tests[6] = { input  = ".test .input *"
           , tokens = { "." , "test" , " " , "." , "input" , " " , "*" }
           , groups = { { id = "test" } , { id = "input" } , ANY }
           , ast    = { orExp = { { andExp = { { id = "test" } , { id = "input" } , ANY } } } }
           , eval   = { { tset = { }                            , result = false }
                      , { tset = { "test" }                     , result = false }
                      , { tset = { "test" , "input" }           , result = true  }
                      , { tset = { "test" , "input" , "extra" } , result = true  }
                      , { tset = { "extra" }                    , result = false }
                      }
           }

tests[7] = { input  = ".test , .input , *"
           , tokens = { "." , "test"  , " , " , "." , "input" , " , " , "*" }
           , groups = { { id = "test" } , OR , { id = "input" } , OR , ANY }
           , ast    = { orExp = { { andExp = { ANY } } , { andExp = { { id = "input" } } } , { andExp = { { id = "test" } } } } }
           , eval   = { { tset = { }                            , result = true }
                      , { tset = { "test" }                     , result = true }
                      , { tset = { "test" , "input" }           , result = true }
                      , { tset = { "test" , "input" , "extra" } , result = true }
                      , { tset = { "extra" }                    , result = true }
                      }
           }

tests[8] = { input  = ".sh"
           , tokens = { "." , "sh" }
           , groups = { { id = "sh" } }
           , ast    = { orExp = { { andExp = { { id = "sh" } } } } }
           , eval   = { { tset = { "sh" }         , result = true  }
                      , { tset = { "sh" , "one" } , result = true  }
                      , { tset = { "one" }        , result = false }
                      , { tset = { }              , result = false }
                      }
           }

tests[9] = { input  = ".sh.one"
           , tokens = { "." , "sh" , "." , "one" }
           , groups = { { id = "sh" } , { id = "one" } }
           , ast    = { orExp = { { andExp = { { id = "sh" } , { id = "one" } } } } }
           , eval   = { { tset = { "sh" }         , result = false }
                      , { tset = { "sh" , "one" } , result = true  }
                      , { tset = { "one" }        , result = false }
                      , { tset = { }              , result = false }
                      }
           }

--- Run Tests
--- ---------

function assert_equal(expected, actual)
    if not deepcompare(expected, actual) then
        print("expected: " .. table.tostring(expected) .. "\nactual:   " .. table.tostring(actual))
        error("test failure!", 1)
    end
end

for i,_ in pairs(tests) do
    print("Test: " .. i)

    print("testing tokenizer...")
    local tokenized = tokenize(tests[i]["input"])
    assert_equal(tests[i]["tokens"], tokenized)

    print("testing token grouper...")
    local grouped = group_tokens(tests[i]["tokens"])
    assert_equal(tests[i]["groups"], grouped)

    print("testing group parser...")
    local parsed = parse_groups(tests[i]["groups"])
    assert_equal(tests[i]["ast"], parsed)

    print("testing main parser...")
    local parsed = parse(tests[i]["input"])
    assert_equal(tests[i]["ast"], parsed)

    print("testing expression evaluator...")
    for _,etest in pairs(tests[i]["eval"]) do
        local evaled = { tset = etest["tset"] , result = eval(tests[i]["ast"], etest["tset"]) }
        assert_equal(etest, evaled)
    end
end
