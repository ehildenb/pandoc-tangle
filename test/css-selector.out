true
false
true
true
false
